File: Sources/ActionRingCore/Profile/ProfileStore.swift
import Foundation

/// Owns the profile database and settings, persists them as JSON, and resolves
/// the active profile via the lookup chain:
///
/// `user(bundleId)` → `builtin(bundleId)` → `user(category)` → `builtin(category)` → `default`
///
/// Used from the main thread (ContextEngine + settings UI); marked
/// `@unchecked Sendable` to satisfy `ProfileProvider` on that contract.
public final class ProfileStore: ProfileProvider, @unchecked Sendable {

    private let store: JSONStore
    private static let profilesFile = "profiles.json"
    private static let defaultFile = "default.json"
    private static let settingsFile = "settings.json"

    public private(set) var userProfiles: [RingProfile]
    public let builtIns: [RingProfile]
    public let categoryDefaults: [AppCategory: RingProfile]
    public private(set) var defaultProfile: RingProfile
    public var settings: AppSettings

    public init(store: JSONStore = JSONStore()) {
        self.store = store
        self.builtIns = BuiltInProfiles.all
        self.categoryDefaults = BuiltInProfiles.categoryDefaults
        self.userProfiles = store.load([RingProfile].self, from: Self.profilesFile) ?? []
        self.defaultProfile = store.load(RingProfile.self, from: Self.defaultFile) ?? .createDefault()
        self.settings = store.load(AppSettings.self, from: Self.settingsFile) ?? AppSettings()
    }

    // MARK: Resolution

    public func resolveProfile(forBundleId bundleId: String?, category: AppCategory) -> RingProfile {
        if let bundleId {
            if let user = userProfiles.first(where: { $0.bundleId == bundleId }) {
                // A user profile overrides the built-in. If the user disabled it,
                // fall through to the category / default chain — the disable was
                // explicit, so don't use the built-in either.
                if user.isEnabled { return user }
            } else if let builtin = builtIns.first(where: { $0.bundleId == bundleId }) {
                return builtin
            }
        }
        if let p = userProfiles.first(where: {
            $0.bundleId == nil && !$0.isDefault && $0.isEnabled && $0.category == category
        }) { return p }
        if let p = categoryDefaults[category] { return p }
        return defaultProfile
    }

    // MARK: Display

    /// All profiles a configuration UI should list: user first, then built-ins
    /// not overridden by a user profile, then the default.
    public func allProfilesForDisplay() -> [RingProfile] {
        let overriddenBundleIds = Set(userProfiles.compactMap { $0.bundleId })
        let visibleBuiltIns = builtIns.filter { builtin in
            guard let id = builtin.bundleId else { return true }
            return !overriddenBundleIds.contains(id)
        }
        return userProfiles + visibleBuiltIns + [defaultProfile]
    }

    // MARK: Mutation

    /// Insert or update a user profile (matched by id), then persist.
    public func upsert(_ profile: RingProfile) {
        var p = profile
        p.touch()
        if let i = userProfiles.firstIndex(where: { $0.id == p.id }) {
            userProfiles[i] = p
        } else {
            userProfiles.append(p)
        }
        persistProfiles()
    }

    public func delete(id: UUID) {
        userProfiles.removeAll { $0.id == id }
        persistProfiles()
    }

    /// Persist an edited profile, routing the universal default to its own file.
    public func save(_ profile: RingProfile) {
        if profile.isDefault { updateDefault(profile) } else { upsert(profile) }
    }

    /// Enable/disable a profile. Built-ins can't be edited in place, so toggling
    /// one materializes an editable user override carrying the new state.
    public func setEnabled(_ profile: RingProfile, _ enabled: Bool) {
        guard !profile.isDefault else { return }   // the default is always on
        // Existing user profile (by id) → update in place.
        if let existing = userProfiles.first(where: { $0.id == profile.id }) {
            var p = existing; p.isEnabled = enabled; upsert(p)
            return
        }
        // A built-in already overridden for the same app → update that override
        // (don't create a second one on repeated toggles).
        if let bundleId = profile.bundleId,
           let existing = userProfiles.first(where: { $0.bundleId == bundleId }) {
            var p = existing; p.isEnabled = enabled; upsert(p)
            return
        }
        // Otherwise materialize a fresh user override carrying the new state.
        var copy = profile
        copy.id = UUID()
        copy.source = .user
        copy.isEnabled = enabled
        upsert(copy)
    }

    /// Returns a user-editable copy of a profile. A built-in becomes a new user
    /// override (fresh id); user profiles and the default are returned as-is.
    public func editableCopy(of profile: RingProfile) -> RingProfile {
        if profile.isDefault { return profile }
        if profile.source == .user, userProfiles.contains(where: { $0.id == profile.id }) {
            return profile
        }
        var copy = profile
        copy.id = UUID()
        copy.source = .user
        return copy
    }

    public func updateDefault(_ profile: RingProfile) {
        var p = profile
        p.isDefault = true
        p.bundleId = nil
        p.touch()
        defaultProfile = p
        store.save(defaultProfile, to: Self.defaultFile)
    }

    public func saveSettings() {
        store.save(settings, to: Self.settingsFile)
    }

    private func persistProfiles() {
        store.save(userProfiles, to: Self.profilesFile)
    }
}

File: windows/src/main/profileStore.js
'use strict';
// ProfileStore — JSON persistence + lookup, port of ProfileStore.swift/JSONStore.swift.
// Files: profiles.json ([RingProfile]), default.json (RingProfile), settings.json
// (AppSettings) under %APPDATA%/ActionRing — same filenames + schema as macOS.
// Writes are atomic (temp + rename). Parse failures are treated as file-absent.

const fs = require('node:fs');
const path = require('node:path');
const {
  uuid, normalizeProfile, normalizeSettings, touch, encodeJSON, DEFAULT_SETTINGS,
} = require('../shared/models');
const { ALL_BUILTINS, categoryDefaults, createDefaultProfile } = require('../shared/builtinProfiles');
const { resolveProfile, allProfilesForDisplay } = require('../shared/resolution');

const PROFILES_FILE = 'profiles.json';
const DEFAULT_FILE = 'default.json';
const SETTINGS_FILE = 'settings.json';

class ProfileStore {
  /** @param {string} directory  e.g. path.join(app.getPath('appData'), 'ActionRing') */
  constructor(directory) {
    this.directory = directory;
    this.builtIns = ALL_BUILTINS;
    this.categoryDefaults = categoryDefaults();
    this.userProfiles = this._load(PROFILES_FILE, (raw) =>
      Array.isArray(raw) ? raw.map(normalizeProfile) : null
    ) ?? [];
    this.defaultProfile = this._load(DEFAULT_FILE, (raw) => {
      const p = normalizeProfile(raw);
      p.isDefault = true; // the universal fallback is always default (mac parity)
      delete p.bundleId;
      return p;
    }) ?? createDefaultProfile();
    this.settings = this._load(SETTINGS_FILE, normalizeSettings) ?? { ...DEFAULT_SETTINGS };
  }

  // ---- disk layer (JSONStore parity: errors → null, atomic writes) ----------

  _load(name, normalize) {
    try {
      const text = fs.readFileSync(path.join(this.directory, name), 'utf8');
      return normalize(JSON.parse(text));
    } catch {
      return null; // missing/corrupt → caller substitutes default
    }
  }

  _save(name, value) {
    try {
      fs.mkdirSync(this.directory, { recursive: true });
      const target = path.join(this.directory, name);
      const tmp = `${target}.${process.pid}.tmp`;
      fs.writeFileSync(tmp, encodeJSON(value), 'utf8');
      fs.renameSync(tmp, target);
      return true;
    } catch {
      return false;
    }
  }

  persistProfiles() { return this._save(PROFILES_FILE, this.userProfiles); }
  persistDefault() { return this._save(DEFAULT_FILE, this.defaultProfile); }
  saveSettings() { return this._save(SETTINGS_FILE, this.settings); }

  // ---- resolution -------------------------------------------------------------

  resolveProfile(appId, category) {
    return resolveProfile(this, appId, category);
  }

  allProfilesForDisplay() {
    return allProfilesForDisplay(this);
  }

  // ---- mutations (ProfileStore.swift semantics) ---------------------------------

  upsert(profile) {
    const copy = structuredClone(profile);
    touch(copy);
    const i = this.userProfiles.findIndex((p) => p.id === copy.id);
    if (i >= 0) this.userProfiles[i] = copy;
    else this.userProfiles.push(copy);
    this.persistProfiles();
  }

  delete(id) {
    this.userProfiles = this.userProfiles.filter((p) => p.id !== id);
    this.persistProfiles();
  }

  /** Routes the universal default to default.json, everything else to profiles.json. */
  save(profile) {
    if (profile.isDefault) this.updateDefault(profile);
    else this.upsert(profile);
  }

  updateDefault(profile) {
    const copy = structuredClone(profile);
    copy.isDefault = true;
    delete copy.bundleId;
    touch(copy);
    this.defaultProfile = copy;
    this.persistDefault();
  }

  /**
   * Toggling a built-in materializes a user override (new id, source user);
   * repeat toggles update the existing override (de-dup by bundleId).
   * The default profile ignores setEnabled.
   */
  setEnabled(profile, enabled) {
    if (profile.isDefault) return;
    const byId = this.userProfiles.find((p) => p.id === profile.id);
    if (byId) {
      const copy = structuredClone(byId);
      copy.isEnabled = enabled;
      this.upsert(copy);
      return;
    }
    if (profile.bundleId != null) {
      const byBundle = this.userProfiles.find((p) => p.bundleId === profile.bundleId);
      if (byBundle) {
        const copy = structuredClone(byBundle);
        copy.isEnabled = enabled;
        this.upsert(copy);
        return;
      }
    }
    const override = structuredClone(profile);
    override.id = uuid();
    override.source = 'user';
    override.isEnabled = enabled;
    this.upsert(override);
  }

  /** Built-in → fresh user copy (new id); user/default returned as-is. */
  editableCopy(profile) {
    if (profile.isDefault) return structuredClone(profile);
    if (profile.source === 'user' && this.userProfiles.some((p) => p.id === profile.id)) {
      return structuredClone(profile);
    }
    const copy = structuredClone(profile);
    copy.id = uuid();
    copy.source = 'user';
    return copy;
  }

  isUserProfile(profile) {
    return profile.source === 'user' && this.userProfiles.some((p) => p.id === profile.id);
  }
}

module.exports = { ProfileStore, PROFILES_FILE, DEFAULT_FILE, SETTINGS_FILE };
